File: C/graphs/DFS.c
#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>


#define MAX 20
#define TRUE 1
#define FALSE 0


int g[MAX][MAX];
int v[MAX];
int n;

void main()
{
    int v1, v2;
    char ans;
    void create();
    void Dfs(int);
    create();
    printf("The Adjacency Matrix for the graph is \n");
    for (v1 = 0; v1 < n; v1++)
    {
        for (v2 = 0; v2 < n; v2++)
            printf("%d", g[v1][v2]);
        printf("\n");
    }
    getchar();
    do
    {
        for (v1 = 0; v1 < n; v1++)
        v[v1] = FALSE;
        printf("Enter the Vertex from which you want to traverse ");
        scanf("%d", &v1);
        if (v1 >= MAX)
            printf("Invalid Vertex\n");
        else
        {
            printf("The Depth First Search of the Graph is\n");
            Dfs(v1);
        }
        printf("\nDo you want to traverse By any other node?");
        ans = getchar();
    } while (ans == 'y');
}
void create()
{
    int ch, v1, v2, flag;
    char ans = 'y';
    printf("\n\t\t This is a program to create a Graph");
    printf("\n\t\t The Display Is In Depth First Manner");
    getch();


    printf("\nEnter no. of nodes");
    scanf("%d", &n);
    for (v1 = 0; v1 < n; v1++)
        for (v2 = 0; v2 < n; v2++)
            g[v1][v2] = FALSE;
    printf("\nEnter no. of nodes");
    scanf("%d", &n);
    printf("\nEnter the vertices no. starting from 0");
    do
    {
        printf("\nEnter the vertices v1 & v2");
        scanf("%d%d", &v1, &v2);
        if (v1 >= n || v2 >= n)
            printf("Invalid Vertex Value\n");
        else
        {
            g[v1][v2] = TRUE;
            g[v2][v1] = TRUE;
        }
        printf("\n\nAdd more edges??(y/n)");
        ans = getchar();
    } while (ans == 'y');
}
void Dfs(int v1)
{
    int v2;
    printf("% d",v1);
        v[v1] = TRUE;
    for (v2 = 0; v2 < MAX; v2++)
        if (g[v1][v2] == TRUE && v[v2] == FALSE)
            Dfs(v2);
}

/*
Sample Case : 
This is a Program To Create a Graph
The Display Is In Depth First Manner
Enter no. of nodes 4
Enter the vertices no. starting from 0
Enter the vertices v1 & v2
0 1
Add more edges??(y/n)y
Enter the vertices v1 & v2
0 2
Add more edges??(y/n)y
Enter the vertices v1 & v2
1 3
Add more edges??(y/n)y
Enter the vertices v1 & v2
2 3
Add more edges??(y/n)
The Adjacency Matrix for the graph is
0 1 1 0
1 0 0 1
1 0 0 1
0 1 1 0
Enter the Vertex from which you want to traverse : 1
The Depth First Search of the Graph is
1
0
2
3
Do you want to traverse from any other node?
*/

File: Python/graphs/dfs.py
class Graph:
    """
    we create an empty adjacency list with values of the nodes as key elements and setting the values as an
    empty list
    """

    elements = []
    """
    Initialing the parents,adjacency list and visited dictionaries.
    """

    def __init__(self, Nodes):
        self.parents = {}
        self.adj_list = {}
        self.node = Nodes
        self.color = {}
        for node in self.node:
            self.adj_list[node] = []
            self.color[node] = "W"
            self.parents[node] = None

    """
    in add_edge function we make and adjacency list of the undirected graph
    """

    def add_edge(self, u, v):
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)

    """
    The main dfs algorithm where we iterate over the neighbours of a node and append them to the element list if 
    it has not been visited before by checking the visited dictionary, also calculate the parent of each node.
    """

    def dfs(self, k):
        # self.visited[u] = 1
        if k not in self.elements:
            self.elements.append(k)
            self.color[k] = "B"
        for v in self.adj_list[k]:
            if self.color[v] == "W":
                self.parents[v] = k
                self.dfs(v)

    """
    The print function print the neighbours of each node, the nodes after iterating over the graph in dfs manner and
    also the parent if each node after executing the dfs function.
    """

    def print(self):
        for node in self.node:
            print(node, "->", self.adj_list[node])
        print("Nodes iterated in the graph:", self.elements)
        print("Parents of each node:", self.parents)


"""
We create the node list along with the object of the class Graph.
specify the nodes and the and the edges in the nodes and edges list before the execution of the program
We call all the functions outside the class using the object created which is 'graph'
"""
nodes = ["A", "B", "C", "D", "E", "F", "G"]
edges = [("A", "B"), ("B", "D"), ("B", "C"), ("E", "F"), ("F", "G")]
"""
The next line creates the instance of the class Graph
"""
graph = Graph(nodes)
color = graph.color
"""
The loop iterates over the edges list and adds them to the adjacency list
"""
for i, j in edges:
    graph.add_edge(i, j)
"""
It iterates over all the nodes and does the depth first search on the tree or the forest
"""
for (m,n) in edges:
    if color[m] == "W":
        graph.dfs(m)
"""
Prints the nodes after iterating over the graph and also the parent of each node  
"""
graph.print()
